src/terminal.rs

```diff
@@ -243,11 +277,19 @@ impl<H: HashT, NI: NodeIdT> Terminal<H, NI> {
         let u_parents = self.unit_store.get(&u_hash).unwrap().parents.clone();
         let mut n_parents_in_dag = NodeCount(0);
         for p_hash in u_parents.into_iter().flatten() {
-            let p = self.unit_store.get(&p_hash).unwrap();
-            if p.status == UnitStatus::InDag {
-                n_parents_in_dag += NodeCount(1);
-            } else {
-                self.add_hash_trigger(&p_hash, u_hash);
+            let maybe_p = self.unit_store.get(&p_hash);
+            // p might not be even in store because u might be a unit with wrong control hash
+            match maybe_p {
+                Some(p) => {
```

Comment:
You may simplify this:
`Some(p) if p.status == UnitStatus::InDag =>n_parents_in_dag += NodeCount(1),`
`_ => self.add_hash_trigger(&p_hash, u_hash),`

Reply:
Right, rewritten!